### PR TITLE
Use SDC library and migrate to TS

### PR DESCRIPTION
**Pull request by tomrf1**
`support-dotcom-components` is our Supporter Revenue messaging/targeting service.
There is [a new package](https://github.com/guardian/support-dotcom-components/pull/605) called @guardian/support-dotcom-components. This is for sharing logic and types with dotcom (frontend + dcr).

Previously we had a very large file, `modules/commercial/contributions-service.js`, for fetching/rendering the epic+banner+header links. In this PR I migrate all of this logic to `modules/support/`, with separate files for epic+banner+header.
I have also migrated this code to typescript, and it makes use of the types exported by the support-dotcom-components library.

This PR migrates to using the new library, and removes the legacy automat dependency.

1. Various types are now imported from the library, e.g. the types of the payloads to be sent to SDC.
2. Uses helper functions for making requests to SDC (e.g. getEpic), with known request/response types.
3. Uses some other helpers, e.g. `incrementWeeklyArticleCount`, which are now defined by the library.

DCR PR: https://github.com/guardian/dotcom-rendering/pull/3819

Tested in CODE:
1. liveblog-epic/banner/header display
2. article counting works
3. article count opt out works
4. liveblog-epic view counting works

### CODE REVIEW

File: static/src/javascripts/projects/common/modules/support/banner.ts [hunk 1]
```diff
@@ -0,0 +1,205 @@
+import { mountDynamic } from '@guardian/automat-modules';
```
Comment (tomrf1) [inline static/src/javascripts/projects/common/modules/support/banner.ts:1]: This file is a TS migration of banner logic from `contributions-service.js`

File: static/src/javascripts/projects/common/modules/support/epic.ts [hunk 2]
```diff
@@ -0,0 +1,142 @@
+import { mountDynamic } from '@guardian/automat-modules';
```
Comment (tomrf1) [inline static/src/javascripts/projects/common/modules/support/epic.ts:1]: This file is a TS migration of epic logic from `contributions-service.js`

File: static/src/javascripts/projects/common/modules/support/header.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
+import { mountDynamic } from '@guardian/automat-modules';
```
Comment (tomrf1) [inline static/src/javascripts/projects/common/modules/support/header.ts:1]: This file is a TS migration of header logic from `contributions-service.js`

File: static/src/javascripts/projects/common/modules/support/supportMessaging.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
+import { onConsentChange } from '@guardian/consent-management-platform';
```
Comment (tomrf1) [inline static/src/javascripts/projects/common/modules/support/supportMessaging.ts:1]: shared code from `contributions-service.js`

File: static/src/javascripts/projects/common/modules/support/banner.ts [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+} from 'common/modules/support/supportMessaging';
+import type { Page } from 'common/modules/support/supportMessaging';
+import userPrefs from 'common/modules/user-prefs';
+import config from 'lib/config';
+import fastdom from 'lib/fastdom-promise';
+import { getCountryCode } from 'lib/geolocation';
+import reportError from 'lib/report-error';
+
+export const NO_RR_BANNER_TIMESTAMP_KEY = 'gu.noRRBannerTimestamp'; // timestamp of when we were last told not to show a RR banner
+const twentyMins = 20 * 60000;
```
Comment (mxdvl) [inline static/src/javascripts/projects/common/modules/support/banner.ts:36]: So many zeroes!

```suggestion
const twentyMins = 20 * 60_000;
```

File: static/src/javascripts/projects/common/modules/support/banner.ts [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+							'subscription-banner : display',
+						);
+					}
+
+					return true;
+				});
+		})
+		.catch((error: any) => {
+			/* eslint-disable @typescript-eslint/restrict-template-expressions -- error log */
+			console.log(`Error importing remote banner: ${error}`);
```
Comment (mxdvl) [inline static/src/javascripts/projects/common/modules/support/banner.ts:100]: I'd suggest using the `@guardian/libs` logger here. It pretty prints logs, and only shows it to relevant teams, even in production.

File: static/src/javascripts/projects/common/modules/support/banner.ts [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+				{},
+				false,
+			);
+			/* eslint-enable @typescript-eslint/restrict-template-expressions */
+			return false;
+		});
+};
+
+const buildBannerPayload = async (): Promise<BannerPayload> => {
+	const page = config.get('page') as Page;
```
Comment (mxdvl) [inline static/src/javascripts/projects/common/modules/support/banner.ts:112]: There's two options to make this type-safe.

Either you use type suggestions for config:

```suggestion
	const page = config.get<Page>('page', {});
```

The preferred approach is to access the window.guardian.page object directly, and updating the type declaration accordingly. This is preferable as it can help us identify exactly what is actually used by our code, and possibly trim the page object in the future.

cc @guardian/dotcom-platform

File: static/src/javascripts/projects/common/modules/support/epic.ts [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+const renderEpic = async (module: ModuleData): Promise<void> => {
+	const Epic = await dynamicImport(module.url, module.name);
+
+	const el = getEpicElement();
+	mountDynamic(el, Epic, { submitComponentEvent, ...module.props }, true);
+};
+
+export const fetchAndRenderEpic = (): Promise<void> => {
+	const contentType = config.get('page.contentType') as string;
```
Comment (mxdvl) [inline static/src/javascripts/projects/common/modules/support/epic.ts:103]: Same as above, I would avoid type assertions wherever possible.

File: static/src/javascripts/projects/common/modules/support/supportMessaging.ts [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+					REQUIRED_CONSENTS_FOR_ARTICLE_COUNT.every(
+						(consent) => tcfv2.consents[consent],
+					);
+
+				if (!hasRequiredConsents) {
+					removeArticleCountsFromLocalStorage();
+				}
+
+				resolve(hasRequiredConsents);
+			}
```
Comment (mxdvl) [inline static/src/javascripts/projects/common/modules/support/supportMessaging.ts:116]: Commercial has recently added a wrapper around getting the initial consent, which ignores a consent signal if it is the cookie banner. This may or may not be relevant for you.

**Review by mxdvl — COMMENTED** [review]
Added some thoughts on TS conversion, the config object, and consent.

File: static/src/javascripts/projects/common/modules/support/banner.ts (hunk 6 above)
Comment (tomrf1) [reply static/src/javascripts/projects/common/modules/support/banner.ts:100]: oh nice! Though I now need to add supporter revenue to the set of teams

File: static/src/javascripts/projects/common/modules/support/banner.ts (hunk 7 above)
Comment (tomrf1) [reply static/src/javascripts/projects/common/modules/support/banner.ts:112]: ahh I didn't realise we have typed `window.guardian.config.page`, that is very useful

File: static/src/javascripts/projects/common/modules/support/supportMessaging.ts (hunk 9 above)
Comment (tomrf1) [reply static/src/javascripts/projects/common/modules/support/supportMessaging.ts:107]: I will leave this logic as it is for this PR, but will take a look

File: static/src/javascripts/projects/common/modules/support/banner.ts (hunk 7 above)
Comment (tomrf1) [reply static/src/javascripts/projects/common/modules/support/banner.ts]: I've updated it to use `window.guardian.config.page`

File: static/src/javascripts/projects/common/modules/support/banner.ts (hunk 6 above)
Comment (tomrf1) [reply static/src/javascripts/projects/common/modules/support/banner.ts]: updating to use `log`, with a new team name!

File: static/src/javascripts/projects/common/modules/support/banner.ts [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+export const renderBanner = (
+	response: ModuleDataResponse,
+): Promise<boolean> => {
+	if (!response.data) {
+		return Promise.resolve(false);
+	}
+	const { module, meta } = response.data;
+
+	return dynamicImport(module.url, module.name)
+		.then((Banner: React.FC<any>) => {
```
Comment (mxdvl) [inline static/src/javascripts/projects/common/modules/support/banner.ts:62]: To avoid an `any`, you can just let the default `{}`.

```suggestion
		.then((Banner: React.FC) => {
```

File: static/src/javascripts/projects/common/modules/support/epic.ts (hunk 8 above)
Comment (mxdvl) [reply static/src/javascripts/projects/common/modules/support/epic.ts:103]: Missed this one?

File: static/src/javascripts/projects/common/modules/support/supportMessaging.ts [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+): Promise<React.FC<any>> => {
+	/* eslint-disable
+			@typescript-eslint/no-unsafe-assignment,
+			@typescript-eslint/no-unsafe-call,
+			@typescript-eslint/no-unsafe-member-access,
+			--
+			dynamic import
+		 */
+	// @ts-expect-error -- see dynamic-import-init.js
+	const component = await window.guardianPolyfilledImport(url);
```
Comment (mxdvl) [inline static/src/javascripts/projects/common/modules/support/supportMessaging.ts:32]: Would it be possible to add types around `window.guardianPolyfilledImport`?

Maybe for a subsequent PR.

File: static/src/javascripts/types/global.d.ts [hunk 12]
```diff
@@ -88,19 +88,22 @@ interface PageConfig extends CommercialPageConfig {
 	nonRefreshableLineItemIds?: number[];
 	section: string;
 	isPaidContent: boolean;
-	isSensitive: boolean;
```
Comment (mxdvl) [inline static/src/javascripts/types/global.d.ts:91]: Why did you remove this key?

**Review by mxdvl — COMMENTED** [review]
I think it’s nearly there. Would you mind removing a few `any`s?

File: static/src/javascripts/projects/common/modules/support/supportMessaging.ts (hunk 11 above)
Comment (tomrf1) [reply static/src/javascripts/projects/common/modules/support/supportMessaging.ts:32]: yes that would be good - dynamic-import-init.js would have to be migrated to TS so I'll leave it for now

File: static/src/javascripts/types/global.d.ts (hunk 12 above)
Comment (tomrf1) [reply static/src/javascripts/types/global.d.ts:91]: ah I should have left a note - it's duplicated on line 81

File: static/src/javascripts/projects/common/modules/support/supportMessaging.ts (hunk 11 above)
Comment (mxdvl) [reply static/src/javascripts/projects/common/modules/support/supportMessaging.ts:32]: Maybe a type declaration is sufficient. See #24596

**Review by mxdvl — APPROVED** [review]
